pylint/checkers/similar.py

```diff
@@ -26,32 +26,315 @@
+        self.first_file: "SuccessiveLinesLimits" = first_file
+        self.second_file: "SuccessiveLinesLimits" = second_file
+        self.effective_cmn_lines_nb: int = effective_cmn_lines_nb
+
+
+# Links the indices ot the starting line in both lineset's stripped lines to
+# the start and end lines in both files
+CplIndexToCplLines_T = Dict["LineSetStartCouple", CplSuccessiveLinesLimits]
+
+
+class LinesChunk:
+    """
+    The LinesChunk object computes and stores the hash of some consecutive stripped lines of a lineset.
+    """
+
+    __slots__ = (
+        "_fileid",  # The name of the file from which the LinesChunk object is generated
+        "_index",  # The index in the stripped lines that is the starting of consecutive lines
+        "_hash",  # The hash of some consecutive lines
+    )
+
+    def __init__(self, fileid: str, num_line: int, *lines: Iterable[str]):
+        self._fileid: str = fileid
+        self._index: Index = Index(num_line)
+        self._hash: int = 0
+        for lin in lines:
+            if lin:
+                self._hash += hash(lin)
+            else:
+                # Avoid matching empty lines
+                self._hash += hash(int(random.random() * 100000))
+
+    def __eq__(self, o: Any) -> bool:
+        if not isinstance(o, LinesChunk):
+            return NotImplemented
+        return self._hash == o._hash
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __repr__(self) -> str:
+        return (
+            f"<LinesChunk object for file {self._fileid} ({self._index}, {self._hash})>"
+        )
+
+    def __str__(self) -> str:
+        return (
+            f"LinesChunk object for file {self._fileid}, starting at line {self._index} \n"
+            f"Hash is {self._hash}"
+        )
+
+
+class SuccessiveLinesLimits:
+    """
+    A class to handle the numbering of begin and end of successive lines.
+
+    :note: Only the end line number can be updated.
+    """
+
+    def __init__(self, start: LineNumber, end: LineNumber):
+        self._start: LineNumber = start
+        self._end: LineNumber = end
+
+    @property
+    def start(self) -> LineNumber:
+        return self._start
+
+    @property
+    def end(self) -> LineNumber:
+        return self._end
+
+    @end.setter
+    def end(self, value: LineNumber):
+        self._end = value
+
+    def __repr__(self) -> str:
+        return f"<SuccessiveLinesLimits <{self._start};{self._end}>>"
+
+
+class LineSetStartCouple(NamedTuple):
+    """
+    Indices in both linesets that mark the beginning of successive lines
+    """
+
+    fst_lineset_index: Index
+    snd_lineset_index: Index
+
+    def __repr__(self) -> str:
+        return (
+            f"<LineSetStartCouple <{self.fst_lineset_index};{self.snd_lineset_index}>>"
+        )
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, LineSetStartCouple):
+            raise NotImplementedError
```

Comment:
```suggestion
            return NotImplemented
```
I think it's a typo as you used this already somewhere else but for reference : https://stackoverflow.com/questions/878943/why-return-notimplemented-instead-of-raising-notimplementederror